WIP: Declarative networkmanager

fkz opened the pull request:
this is inspired by #12219. Still work in progress.
As this is my first usage of the submodule system, any comments would be appreciated :smile:

akavel committed:
network-manager: declarative config

fkz committed:
networkmanager: declarative configuration

copumpkin commented on nixos/modules/services/networking/networkmanager-connections/wifi.nix:
```diff
@@ -0,0 +1,49 @@
+  options = {
+    password = mkOption {
+      type = types.nullOr types.str;
+      description = "The WPA2 password of the connection";
```
I'd make it very clear in the description that this will end up world-readable in the store

fkz replied:
changed, thanks!

heydojo commented:
Could gnupg not be used to prevent plain text world readable passwords please?
https://wiki.archlinux.org/index.php/GnuPG#Encrypt_a_password
Storing a password in plain text completely defeats the object behind creating one IMHO.

copumpkin commented:
@heydojo thoughts on @edolstra's https://github.com/edolstra/nixpkgs/commit/4c8212069429bf9fb959e00ce8d9345ac7cb7ff0?

copumpkin commented:
The main reason I keep bringing this up is that I'd like us to have some option for secrets in the store, and this seems like a lightweight approach that avoids big changes with a deep impact to the Nix codebase.

heydojo commented:
@copumpkin that certainly looks like the ticket :) :+1:

fkz commented:
that looks reasonable. We have to merge https://github.com/edolstra/nix/commit/6b7003672ff8d73d2a1f3b6709b216b3d0a7ad9c (add encrypt primop) into nix for that first.

fkz commented:
@edolstra could we merge edolstra/nix@6b70036 (add encrypt primop) into nix? I think this would be a really useful primop.

@heydojo I don't think the gnupg approach would work, since we still need to create a derivation which encrypts the password; hence it's world-readable. So I don't think there's a way to archieve this without adding the mentioned primop (other than implementing it in nix... which is probably far too slow)

copumpkin commented:
I think that the primop should be added with a "provisional" sort of documentation, saying basically "we'd like to experiment with ways to provide secret store data, and this is a hypothesis. Try it out and provide feedback, but we're not committed to the API"

fkz commented:
I just realized that the encrypt primop will generate a different string at each evaluation time because every time a new nonce is chosen. This is unfortunate but probably necessary.
We could set the nonce as a possible third argument but then we'd have to be sure to use a different nonce at all cases (user-wise). Still thinking at ways how this nonce can somehow be 'cached' (somehow by location).
Anyway, like this we have to be sure to only use lightweight derivations that depend on passwords - which should normally be the case.

fkz committed:
move assertion implementation into assertions.nix (where declaration already is)

This was previously in top-level.nix, and so could only be used in the nixos module system.
Like this we can also use assertions (which are a generally useful feature) in other (sub)module systems.

grahamc commented:
I'm pretty interested in this feature. Are you still interested in implementing it?

fkz commented:
@grahamc hey, sorry for the delay, I plan to work on it at http://www.meetup.com/de-DE/Berlin-NixOS-Meetup/events/231497025/

ttuegel commented:
I am closing this PR because it never progressed past WIP status and it now has merge conflicts. Please open a new PR if you want to take up work on this again.
